src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +27,141 @@ class DictionarySegment : public BaseSegment {
+    }
+    return (ValueID)(forward_iterator - _dictionary->begin());
+  }
 
   // same as upper_bound(T), but accepts an AllTypeVariant
-  ValueID upper_bound(const AllTypeVariant& value) const;
+  ValueID upper_bound(const AllTypeVariant& value) const { return upper_bound(value); }
 
   // return the number of unique_values (dictionary entries)
-  size_t unique_values_count() const;
+  size_t unique_values_count() const { return _dictionary->size(); }
 
   // return the number of entries
-  ChunkOffset size() const override;
+  ChunkOffset size() const override { return _attribute_vector->size(); }
 
   // returns the calculated memory usage
-  size_t estimate_memory_usage() const final;
+  size_t estimate_memory_usage() const final {
+    return _dictionary->size() * sizeof(T) + _attribute_vector->size() * _attribute_vector->width();
+  }
 
  protected:
   std::shared_ptr<std::vector<T>> _dictionary;
   std::shared_ptr<BaseAttributeVector> _attribute_vector;
-};
 
+  std::shared_ptr<std::vector<T>> _initialize_dictionary(const std::shared_ptr<ValueSegment<T>> value_segment) {
+    std::set<T> distinct_values;
+    for (auto value : value_segment->values()) {
+      distinct_values.insert(value);
+    }
+    return std::make_shared<std::vector<T>>(distinct_values.begin(), distinct_values.end());
+  }
+
+  int _get_minimal_number_of_bits_for_dictionary_size(size_t size) {
+    if (size <= std::numeric_limits<uint8_t>::max()) {
+      return 8;
+    } else if (size <= std::numeric_limits<uint16_t>::max()) {
+      return 16;
+    } else if (size <= std::numeric_limits<uint32_t>::max()) {
+      return 32;
+    } else {
+      return 64;
+    }
+  }
+
+  std::shared_ptr<BaseAttributeVector> _create_attribute_vector_with_type(
+      int minimal_bytes, std::shared_ptr<ValueSegment<T>> value_segment, std::shared_ptr<std::vector<T>> dictionary) {
+    std::shared_ptr<BaseAttributeVector> result;
+    switch (minimal_bytes) {
+      case 8:
+        result = _initialize_attribute_vector<uint8_t>(dictionary, value_segment);
+        break;
+      case 16:
+        result = _initialize_attribute_vector<uint16_t>(dictionary, value_segment);
+        break;
+      case 32:
+        result = _initialize_attribute_vector<uint32_t>(dictionary, value_segment);
+        break;
+      default:
+        result = _initialize_attribute_vector<uint64_t>(dictionary, value_segment);
+    }
+    return result;
+  }
+
+  template <typename uintX_t>
+  std::shared_ptr<BaseAttributeVector> _initialize_attribute_vector(std::shared_ptr<std::vector<T>> dictionary,
+                                                                    std::shared_ptr<ValueSegment<T>> value_segment) {
+    auto fixed_attribute_vector = std::make_shared<FixedSizeAttributeVector<uintX_t>>(value_segment->size());
+    std::shared_ptr<BaseAttributeVector> attribute_vector =
+        std::dynamic_pointer_cast<BaseAttributeVector>(fixed_attribute_vector);
+
+    std::map<T, uintX_t> values_to_index;
```

Comment:
Using the map is fine but it consumes additional memory compared to using `lower_bound` on the dictionary.